### PR TITLE
feat(topolotree): new implementation and Hydro Deploy setup

### PR DESCRIPTION
**Pull request by shadaj**
feat(topolotree): new implementation and Hydro Deploy setup

--

Co-authored-by: zzlk <2418897+zzlk@users.noreply.github.com>
Co-authored-by: Saikrishna Achalla <saikrishnaachalla@Saikrishnas-MacBook-Pro.local>

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/909).
* #914
* #910
* __->__ #909

### CODE REVIEW

**Comment by shadaj** [comment]
This includes many quality-of-life improvements to Hydro Deploy as well, so probably good to merge this sooner than later.

**Review by Saikrishna-Achalla — COMMENTED** [review]
In `topolotree/topolotree_latency.hydro.py`, this is still binary tree support right? Otherwise lgtm

File: hydro_cli/src/core/hydroflow_crate/build.rs [hunk 1]
```diff
@@ -16,6 +16,7 @@ type CacheKey = (
     PathBuf,
     Option<String>,
     Option<String>,
+    Option<String>,
     HostTargetType,
     Option<Vec<String>>,
 );
```
Comment (MingweiSamuel) [inline hydro_cli/src/core/hydroflow_crate/build.rs:22]: this really should be a struct with named fields

File: hydro_cli/src/core/localhost.rs [hunk 2]
```diff
@@ -69,17 +77,39 @@ impl LaunchedBinary for LaunchedLocalhostBinary {
...
 pub fn create_broadcast<T: AsyncRead + Send + Unpin + 'static>(
     source: T,
     default: impl Fn(String) + Send + 'static,
-) -> Arc<RwLock<Vec<Sender<String>>>> {
+) -> CLIAndMain {
+    let cli_receivers = Arc::new(RwLock::new(Vec::<Sender<String>>::new()));
     let receivers = Arc::new(RwLock::new(Vec::<Sender<String>>::new()));
+
+    let weak_cli_receivers = Arc::downgrade(&cli_receivers);
     let weak_receivers = Arc::downgrade(&receivers);
```
Comment (MingweiSamuel) [inline hydro_cli/src/core/localhost.rs:93]: Add a comment - whats the difference between CLI receivers and main receivers

File: topolotree/src/tests.rs [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+//     let input2 = (1, Payload {timestamp:4, data:2});
+//     let output1: (u32, Payload) = (2, Payload {timestamp:1, data:7});
+//     let output2: (u32, Payload) = (3, Payload {timestamp:1, data:7});
+//
+// updates from multiple sources test
+//     let input1 = (1, Payload {timestamp:5, data:7});
+//     let input2 = (2, Payload {timestamp:4, data:2});
+//     let output1: (u32, Payload) = (1, Payload {timestamp:2, data:2});
+//     let output2: (u32, Payload) = (2, Payload {timestamp:2, data:7});
+//     let output3: (u32, Payload) = (3, Payload {timestamp:2, data:9});
```
Comment (MingweiSamuel) [inline topolotree/src/tests.rs:424]: What are these from/for?

**Review by MingweiSamuel — APPROVED** [review]
Looks good

**Pull request merged**